wizard/src/utils/obs_generation.ts

```diff
@@ -0,0 +1,148 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import _ from "lodash";
+
+import {
+  Column,
+  CsvData,
+  MappedThing,
+  Mapping,
+  MappingType,
+  MappingVal,
+  Observation,
+  RowObservations,
+} from "../types";
+
+// Helper maps used while generating observaitons.
+interface ObsGenMaps {
+  // Column Index -> mapped thing found in header for VALUE.
+  valCol2Hdr: Map<number, MappedThing>;
+  // Mapped thing -> column index (for non-VALUE)
+  thing2Col: Map<MappedThing, number>;
+  // Mapped thing -> constant value (for non-VALUE)
+  thing2Const: Map<MappedThing, string>;
+}
+
+function hasRequiredProps(obs: Observation): boolean {
+  for (const mthing of [
+    MappedThing.PLACE,
+    MappedThing.STAT_VAR,
+    MappedThing.DATE,
+    MappedThing.VALUE,
+  ]) {
+    if (!obs.has(mthing)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function generateObservationsInRow(
+  row: Array<string>,
+  header: Array<Column>,
+  obsGenMaps: ObsGenMaps
+): Array<Observation> {
+  const obsList = new Array<Observation>();
+  for (const valColIdx of Array.from(obsGenMaps.valCol2Hdr.keys())) {
+    const hdrThing = obsGenMaps.valCol2Hdr.get(valColIdx);
+    const obs: Observation = new Map();
+    // TODO: Consider if we want to do some cleaning and checking for non-numeric value.
+    if (!_.isEmpty(row[valColIdx])) {
+      obs.set(MappedThing.VALUE, row[valColIdx]);
+    }
+    // If this is a COLUMN_HEADER case, get the corresponding header value.
+    if (hdrThing !== MappedThing.VALUE) {
+      obs.set(hdrThing, header[valColIdx].header);
+    }
+    for (const mthing of Array.from(obsGenMaps.thing2Col.keys())) {
+      const colIdx = obsGenMaps.thing2Col.get(mthing);
+      if (!_.isEmpty(row[colIdx])) {
+        obs.set(mthing, row[colIdx]);
+      }
+    }
+    for (const mthing of Array.from(obsGenMaps.thing2Const.keys())) {
+      obs.set(mthing, obsGenMaps.thing2Const.get(mthing));
+    }
+    if (hasRequiredProps(obs)) {
+      obsList.push(obs);
+    }
+  }
+  return obsList;
+}
+
+/**
+ * Given a mapping and CSV Data computes the observations for every row
+ * corresponding to |rowsForDisplay|.
```

Comment:
nit: this function comment does not seem to follow the function comment format with @param and @returns etc

Reply:
Thanks for the reminder :)